### PR TITLE
refactor: separate data into staticData, encodedLengths, dynamicData in getRecord

### PR DESCRIPTION
**Pull request by dk1a**
fixes https://github.com/latticexyz/mud/issues/1518

### CODE REVIEW

File: e2e/packages/contracts/src/codegen/tables/Vector.sol [hunk 1]
```diff
@@ -277,13 +304,13 @@ library Vector {
   }
 
   /** Tightly pack full data using this table's field layout */
-  function encode(int32 x, int32 y) internal pure returns (bytes memory) {
+  function encode(int32 x, int32 y) internal pure returns (bytes memory, PackedCounter, bytes memory) {
```
Comment (holic) [inline e2e/packages/contracts/src/codegen/tables/Vector.sol:307]: does it help at all gas wise to name these return values and use them in the function body?

```solidity
function encode(int32 x, int32 y) internal pure returns (bytes memory _staticData, PackedCounter, bytes memory) {
  _staticData = encodeStatic(x, y);

```

File: packages/store/src/StoreCore.sol [hunk 2]
```diff
@@ -521,54 +521,31 @@ library StoreCore {
     bytes32 tableId,
     bytes32[] memory keyTuple,
     FieldLayout fieldLayout
-  ) internal view returns (bytes memory) {
+  ) internal view returns (bytes memory staticData, PackedCounter encodedLengths, bytes memory dynamicData) {
```
Comment (holic) [inline packages/store/src/StoreCore.sol:524]: since we're now passing around this tuple a lot, do you think we'd get any value from encoding this as a struct? or would that make it less convenient?

```solidity
struct ValueData {
  bytes staticData;
  PackedCounter encodedLengths;
  bytes dynamicData;
}
```

File: packages/store/ts/codegen/record.ts [hunk 3]
```diff
@@ -115,69 +120,95 @@ function renderDecodeFunction({ structName, fields, staticFields, dynamicFields
...
-              } else {
-                return `
-                  _start = _end;
-                  unchecked {
-                    _end += _encodedLengths.atIndex(${index});
-                  }
-                  ${fieldNamePrefix}${field.name} = ${renderDecodeDynamicFieldPartial(field)};
-                `;
+    result += `
+    /**
```
Comment (holic) [inline packages/store/ts/codegen/record.ts:146]: nit: can we indent this one level (and all the backticks below) for readability?

File: .changeset/hip-tables-check.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+---
+"@latticexyz/store": major
+"@latticexyz/world": major
```
Comment (holic) [inline .changeset/hip-tables-check.md:3]:
```suggestion
"@latticexyz/world": patch
```
As far as I can tell, this didn't break any world APIs, this is just a bump of the store dep?

File: packages/store/src/StoreCore.sol (hunk 2 above)
Comment (holic) [reply packages/store/src/StoreCore.sol:524]: I see a lot of folks asking about stack too deep errors, is this anything to consider?

File: .changeset/hip-tables-check.md (hunk 4 above)
Comment (dk1a) [reply .changeset/hip-tables-check.md]: yeah, keyswithvalue got an internal breaking change, but that's not api

File: packages/store/src/StoreCore.sol (hunk 2 above)
Comment (dk1a) [reply packages/store/src/StoreCore.sol:524]: yes, it would reduce vars from 3 to 1

File: .changeset/hip-tables-check.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
+---
+"@latticexyz/store": major
+"@latticexyz/world": patch
```
Comment (holic) [inline .changeset/hip-tables-check.md:3]: this got dismissed with applying the change but I am wondering about moving this to its own changeset

otherwise, the world changelog will have info about store internals changes (which aren't super relevant if you're using world)

Comment (alvrs) [reply .changeset/hip-tables-check.md:3]: Since `StoreRead`'s API was changed, also the `World` API changed, so i'd say this is a `major` change for world?

Comment (holic) [reply .changeset/hip-tables-check.md:3]: Ohh right, very true

File: packages/store/src/StoreCore.sol (hunk 2 above)
Comment (alvrs) [reply packages/store/src/StoreCore.sol:524]: might be worth it to support more variables in generated libraries? Since the functions are not used a lot outside codegen I think the slightly decreased dev ex of having to import a struct for the return data is fine

> edit: nevermind, talked myself out of this idea

curious why?

Comment (alvrs) [reply packages/store/src/StoreCore.sol:524]: the main advantage of using the struct is in the `set` function, where I believe we'd currently run into "stack too deep" issues for tables where num keys + num fields is greater than 7 (because the set function has 5 additional variables besides the key+fields). That might be worked around if we make the version of setter that uses the table's data struct a "first class citizen" instead of internally calling the version with arguments.

In any case, i think we can do this as a follow up since it involves refactors to code unrelated to this PR. Opened an issue to track here: #1535

**Review by alvrs — DISMISSED** [review]
Looks good! Let's do the struct refactor in a follow up (tracked in #1535 and #1536). Only nits left are the indentation mentioned by @holic and changing the changeset back to major.

File: packages/store/ts/codegen/record.ts (hunk 3 above)
Comment (dk1a) [reply packages/store/ts/codegen/record.ts:146]: sure, I went with this style for all renderers so just sticking to it here. If we wanna change it I'd rather make a separate PR and change it everywhere

Comment (holic) [reply packages/store/ts/codegen/record.ts:146]: we already did this: https://github.com/latticexyz/mud/pull/1474

**Review by alvrs — APPROVED** [review]

File: e2e/packages/contracts/src/codegen/tables/Vector.sol (hunk 1 above)
Comment (dk1a) [reply e2e/packages/contracts/src/codegen/tables/Vector.sol:307]: tested on `Callbacks` and `Vector2` but it doesn't seem to make a difference. Intuitively I'd also assume that this shouldn't affect bytecode
